### PR TITLE
plugins/lsp/language-servers: nil_ls add nix options

### PR DESCRIPTION
**Pull request by khaneliman**
Adding some options that were being ignored when I tried to add to config for nil_ls. 
https://github.com/oxalica/nil/blob/main/docs/configuration.md?plain=1

### CODE REVIEW

File: plugins/lsp/language-servers/default.nix [hunk 1]
```diff
@@ -389,10 +389,64 @@ let
...
+              '';
+            };
+          };
+        };
       };
       settings = cfg: {
         nil = {
-          inherit (cfg) formatting diagnostics;
+          inherit (cfg) formatting diagnostics nix;
         };
```
Comment (MattSturgeon) [inline plugins/lsp/language-servers/default.nix:450]: > options that were being ignored when I tried to add to config for `nil_ls`

Should this just be `nil = cfg;`, is inheriting specific options the reason why this didn't work as-per RFC42?

@GaetanLepage @traxys

Comment (khaneliman) [reply plugins/lsp/language-servers/default.nix]: Makes sense, I tested that built my config still and supports new attribute sets not defined.

**Comment by khaneliman** [comment]
Sorry for false push, meant to test locally before pushing that commit...

**Comment by khaneliman** [comment]
Alright, seems to work same as before with new file for settings.

**Review by MattSturgeon — APPROVED** [review]
LGTM. It'd be great if this added a test (similar to the [`nixd` test](https://github.com/khaneliman/nixvim/blob/nil/tests/test-sources/plugins/lsp/language-servers/nixd.nix)) with example and/or defaults, but so few language servers have tests I don't think this is a blocker.

**Review by GaetanLepage — CHANGES_REQUESTED** [review]

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+#  - https://github.com/oxalica/nil/blob/main/docs/configuration.md?plain=1#
+with lib; {
+  formatting = {
+    command = helpers.defaultNullOpts.mkListOf types.str "null" ''
+      External formatter command (with arguments).
+      It should accepts file content in stdin and print the formatted code into stdout.
+    '';
+  };
+  diagnostics = {
+    ignored = helpers.defaultNullOpts.mkListOf types.str "null" ''
```
Comment (MattSturgeon) [inline plugins/lsp/language-servers/nil-ls-settings.nix:15]:
```suggestion
    ignored = helpers.defaultNullOpts.mkListOf types.str "[]" ''
```

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+      Ignored diagnostic kinds.
+      The kind identifier is a snake_cased_string usually shown together
+      with the diagnostic message.
+    '';
+    excludedFiles = helpers.defaultNullOpts.mkListOf types.str "null" ''
```
Comment (MattSturgeon) [inline plugins/lsp/language-servers/nil-ls-settings.nix:20]:
```suggestion
    excludedFiles = helpers.defaultNullOpts.mkListOf types.str "[]" ''
```

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+      Files to exclude from showing diagnostics. Useful for generated files.
+      It accepts an array of paths. Relative paths are joint to the workspace root.
+      Glob patterns are currently not supported.
+    '';
+  };
+  nix = {
+    binary = helpers.defaultNullOpts.mkStr "null" ''
```
Comment (MattSturgeon) [inline plugins/lsp/language-servers/nil-ls-settings.nix:27]:
```suggestion
    binary = helpers.defaultNullOpts.mkStr "nix" ''
```

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+      The path to the `nix` binary.
+    '';
+    maxMemoryMB = helpers.defaultNullOpts.mkInt "null" ''
```
Comment (MattSturgeon) [inline plugins/lsp/language-servers/nil-ls-settings.nix:30]:
```suggestion
    maxMemoryMB = helpers.defaultNullOpts.mkInt 2560 ''
```

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      about 2GiB memory.
+    '';
+    flake = {
+      autoArchive = helpers.defaultNullOpts.mkBool "null" ''
+        Auto-archiving behavior which may use network.
+         - null: Ask every time.
+         - true: Automatically run `nix flake archive` when necessary.
+         - false: Do not archive. Only load inputs that are already on disk.
+      '';
+      autoEvalInputs = helpers.defaultNullOpts.mkBool "null" ''
```
Comment (MattSturgeon) [inline plugins/lsp/language-servers/nil-ls-settings.nix:45]:
```suggestion
      autoEvalInputs = helpers.defaultNullOpts.mkBool false ''
```

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        Whether to auto-eval flake inputs.
+        The evaluation result is used to improve completion, but may cost
+        lots of time and/or memory.
+      '';
+      nixpkgsInputName = helpers.defaultNullOpts.mkStr "null" ''
```
Comment (MattSturgeon) [inline plugins/lsp/language-servers/nil-ls-settings.nix:50]:
```suggestion
      nixpkgsInputName = helpers.defaultNullOpts.mkStr "nixpkgs" ''
```

**Review by MattSturgeon — CHANGES_REQUESTED** [review]
The default passed to `helpers.defaultNullOpts` is the _plugin default_, it will always produce an option with the _nix default_ set to `null`.

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 8]
```diff
@@ -0,0 +1,60 @@
+{
+  lib,
+  helpers,
+}:
+# Options:
+#  - https://github.com/oxalica/nil/blob/main/docs/configuration.md?plain=1#
+with lib; {
+  formatting = {
+    command = helpers.defaultNullOpts.mkListOf types.str null ''
```
Comment (MattSturgeon) [inline plugins/lsp/language-servers/nil-ls-settings.nix:9]: `mkListOf str` produces a list of `maybeRaw str`, which is different from the `listOf str` we had before.

Would something like this be more correct @GaetanLepage?

```nix
mkNullable (listOf str) "null" "<description-here>"
```

Or should this type accept `__raw` lua strings @khaneliman?

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+      External formatter command (with arguments).
+      It should accepts file content in stdin and print the formatted code into stdout.
+    '';
+  };
+  diagnostics = {
+    ignored = helpers.defaultNullOpts.mkListOf types.str [] ''
```
Comment (MattSturgeon) [inline plugins/lsp/language-servers/nil-ls-settings.nix:15]: Same

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+      Ignored diagnostic kinds.
+      The kind identifier is a snake_cased_string usually shown together
+      with the diagnostic message.
+    '';
+    excludedFiles = helpers.defaultNullOpts.mkListOf types.str [] ''
```
Comment (MattSturgeon) [inline plugins/lsp/language-servers/nil-ls-settings.nix:20]: Same

File: plugins/lsp/language-servers/nil-ls-settings.nix (hunk 8 above)
Comment (khaneliman) [reply plugins/lsp/language-servers/nil-ls-settings.nix:9]: I believe it'd just be the executable path or name.

**Comment by MattSturgeon** [comment]
CI is failing on formatting, if you want you can install the pre-commit hook by entering the devshell at least once (you can exit again and use your normal shell if you like, entering once is enough to install the hook). Enter using `nix develop`.

`nix fmt` should fix the formatting, at least.

**Comment by khaneliman** [comment]
> CI is failing on formatting, if you want you can install the pre-commit hook by entering the devshell at least once (you can exit again and use your normal shell if you like, entering once is enough to install the hook). Enter using `nix develop`.
> 
> Installing the hook will check the formatting of modified files before each commit.
> 
> `nix fmt` should fix the formatting, at least.

The pre-commit hook was running on my commits. I'll let it format the full file, but I tried to keep changes to just my code.

File: plugins/lsp/language-servers/default.nix [hunk 11]
```diff
@@ -275,19 +274,18 @@ let
...
       # Use the old name of the lua LS if the user is on a stable branch of nixpkgs
       # Rename occurred here: https://github.com/NixOS/nixpkgs/pull/215057
       package =
-        if (hasAttr "lua-language-server" pkgs) then
-          pkgs.lua-language-server
-        else
-          pkgs.sumneko-lua-language-server;
+        if (hasAttr "lua-language-server" pkgs)
+        then pkgs.lua-language-server
+        else pkgs.sumneko-lua-language-server;
```
Comment (MattSturgeon) [inline plugins/lsp/language-servers/default.nix:288]: Strange that unrelated code needs to be formatted...

Was this definitely `nix fmt` and not some additional formatter you have in your editor/workflow? The result looks like [Alejandra](https://github.com/kamadorueda/alejandra).

If this is coming from `nix fmt`, maybe rebase into a separate commits; one doing `nix fmt` and another adding this PR's changes.

Comment (MattSturgeon) [reply plugins/lsp/language-servers/default.nix]: > Looks like the hook is fighting the nix fmt..

Interesting

Comment (khaneliman) [reply plugins/lsp/language-servers/default.nix]: I had to direnv reload so it would reinstall the pre-commit-hook. It must have changed.

File: plugins/lsp/language-servers/nil-ls-settings.nix (hunk 8 above)
Comment (khaneliman) [reply plugins/lsp/language-servers/nil-ls-settings.nix:7]: Gonna be hitting the road for a few hours. If this needs to change I can address later.

File: plugins/lsp/language-servers/default.nix (hunk 11 above)
Comment (MattSturgeon) [reply plugins/lsp/language-servers/default.nix]: I can't see why there'd be an issue; The only difference is `nix fmt` is wrapped with `treefmt` while `pre-commit` is directly using `statix` and `nixfmt-rfc-style`, which `treefmt` is _also_ configured to use... ([here](https://github.com/nix-community/nixvim/blob/9d858de2e9ab136d1c53d92af62fed8fccf492ab/flake-modules/dev/default.nix#L12-L36))

Comment (MattSturgeon) [reply plugins/lsp/language-servers/default.nix]: > Was the repo using Alejandra before and switched to new nixfmt-rfc-style?

Yeah, it was

> I had to direnv reload so it would reinstall the pre-commit-hook.

Makes sense, didn't consider you might have had an old hook installed :smile:

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 12]
```diff
@@ -0,0 +1,58 @@
+{ lib, helpers }:
+# Options:
+#  - https://github.com/oxalica/nil/blob/main/docs/configuration.md?plain=1#
+with lib;
+{
+  formatting = {
+    command = helpers.defaultNullOpts.mkNullable (types.listOf types.str) null ''
```
Comment (MattSturgeon) [inline plugins/lsp/language-servers/nil-ls-settings.nix:7]: I was under the impression `mkNullable` needed the plugin-default to be a string (`"null"`), but it looks like it gets passed through `toString`.

I guess we only need to pass in a string when we want to enforce formatting or display nested elements. 🤔

**Review by MattSturgeon — APPROVED** [review]
CI is passing and the diff looks good 👍 

Thanks for your patience!

File: plugins/lsp/language-servers/nil-ls-settings.nix (hunk 12 above)
Comment (GaetanLepage) [reply plugins/lsp/language-servers/nil-ls-settings.nix:7]: Hmmm, we should be careful here.
Indeed, `toString null` is `""` and not `"null"`.

Comment (MattSturgeon) [reply plugins/lsp/language-servers/nil-ls-settings.nix:7]: Off-topic, but maybe `mkNullable` should use a (bespoke or third-party) pretty-stringify instead of `builtins.toString` to allow us to use actual `nix` and strings interchangeably?

For now, @khaneliman, could you switch to passing string defaults to `mkNullable` matching how it's done elsewhere?

Comment (GaetanLepage) [reply plugins/lsp/language-servers/nil-ls-settings.nix:7]: > Off-topic, but maybe `mkNullable` should use a (bespoke or third-party) pretty-stringify instead of `builtins.toString` to allow us to use actual `nix` and strings interchangeably?

I agree, this could be nice !

Comment (khaneliman) [reply plugins/lsp/language-servers/nil-ls-settings.nix]: Sounds good, updated to use the "null" string instead of null type.

File: plugins/lsp/language-servers/nil-ls-settings.nix (hunk 8 above)
Comment (MattSturgeon) [reply plugins/lsp/language-servers/nil-ls-settings.nix]: I'd like to run this by @traxys as his recent comments on one of my PRs are making me second guess this feedback.

Raw strings are really just lua code; i.e. no quotes. So a user might want to use a raw string here to reference some other lua string-variable defined elsewhere.

_I guess_ the only time we'd want to avoid them is when the value _won't_ end up directly in lua code, or we're going to do some string manipulation on it (and assume it is a string literal)

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 13]
```diff
@@ -0,0 +1,58 @@
+{ lib, helpers }:
+# Options:
+#  - https://github.com/oxalica/nil/blob/main/docs/configuration.md?plain=1#
+with lib;
+{
+  formatting = {
+    command = helpers.defaultNullOpts.mkNullable (types.listOf types.str) "null" ''
```
Comment (GaetanLepage) [inline plugins/lsp/language-servers/nil-ls-settings.nix:7]:
```suggestion
    command = helpers.defaultNullOpts.mkListOf types.str "null" ''
```
Use `mkListOf` where possible.

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+      External formatter command (with arguments).
+      It should accepts file content in stdin and print the formatted code into stdout.
+    '';
+  };
+  diagnostics = {
```
Comment (GaetanLepage) [inline plugins/lsp/language-servers/nil-ls-settings.nix:12]:
```suggestion
  };

  diagnostics = {
```

Please, add a blank line between each option declaration.

**Review by GaetanLepage — CHANGES_REQUESTED** [review]
A few minor requests.
Thanks for your patience to address all of those comments.

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+    '';
+  };
+
+  diagnostics = {
+    ignored = helpers.defaultNullOpts.mkListOf types.str [ ] ''
+      Ignored diagnostic kinds.
+      The kind identifier is a snake_cased_string usually shown together
+      with the diagnostic message.
+    '';
+    excludedFiles = helpers.defaultNullOpts.mkListOf types.str [ ] ''
```
Comment (GaetanLepage) [inline plugins/lsp/language-servers/nil-ls-settings.nix:19]:
```suggestion
    '';

    excludedFiles = helpers.defaultNullOpts.mkListOf types.str [ ] ''
```

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+      It accepts an array of paths. Relative paths are joint to the workspace root.
+      Glob patterns are currently not supported.
+    '';
+  };
+
+  nix = {
+    binary = helpers.defaultNullOpts.mkStr "nix" ''
+      The path to the `nix` binary.
+    '';
+    maxMemoryMB = helpers.defaultNullOpts.mkInt 2560 ''
```
Comment (GaetanLepage) [inline plugins/lsp/language-servers/nil-ls-settings.nix:30]:
```suggestion
    '';

    maxMemoryMB = helpers.defaultNullOpts.mkInt 2560 ''
```

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+      The heap memory limit in MiB for `nix` evaluation.
+      Currently it only applies to flake evaluation when `autoEvalInputs` is
+      enabled, and only works for Linux. Other `nix` invocations may be also
+      applied in the future. `null` means no limit.
+      As a reference, `nix flake show --legacy nixpkgs` usually requires
+      about 2GiB memory.
+    '';
+    flake = {
```
Comment (GaetanLepage) [inline plugins/lsp/language-servers/nil-ls-settings.nix:38]:
```suggestion
    '';

    flake = {
```

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+      autoArchive = helpers.defaultNullOpts.mkBool false ''
+        Auto-archiving behavior which may use network.
+         - null: Ask every time.
+         - true: Automatically run `nix flake archive` when necessary.
+         - false: Do not archive. Only load inputs that are already on disk.
+      '';
+      autoEvalInputs = helpers.defaultNullOpts.mkBool false ''
```
Comment (GaetanLepage) [inline plugins/lsp/language-servers/nil-ls-settings.nix:45]:
```suggestion
      '';

      autoEvalInputs = helpers.defaultNullOpts.mkBool false ''
```

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+        Whether to auto-eval flake inputs.
+        The evaluation result is used to improve completion, but may cost
+        lots of time and/or memory.
+      '';
+      nixpkgsInputName = helpers.defaultNullOpts.mkStr "nixpkgs" ''
```
Comment (GaetanLepage) [inline plugins/lsp/language-servers/nil-ls-settings.nix:50]:
```suggestion
      '';

      nixpkgsInputName = helpers.defaultNullOpts.mkStr "nixpkgs" ''
```

File: tests/test-sources/plugins/lsp/language-servers/nil-ls.nix [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+        enable = true;
+
+        settings = {
+          diagnostics = {
+            ignored = [
+              "unused_binding"
+              "unused_with"
+            ];
+
+            excludedFiles = [ "Cargo.nix" ];
```
Comment (GaetanLepage) [inline tests/test-sources/plugins/lsp/language-servers/nil-ls.nix:16]:
```suggestion
            ];
            excludedFiles = [ "Cargo.nix" ];
```

File: tests/test-sources/plugins/lsp/language-servers/nil-ls.nix [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+          };
+          formatting = {
+            command = [ "nixfmt" ];
+          };
+          nix = {
+            binary = "nix";
+            maxMemoryMB = 2048;
+
+            flake = {
```
Comment (GaetanLepage) [inline tests/test-sources/plugins/lsp/language-servers/nil-ls.nix:25]:
```suggestion
            maxMemoryMB = 2048;
            flake = {
```

**Review by GaetanLepage — CHANGES_REQUESTED** [review]

File: plugins/lsp/language-servers/nil-ls-settings.nix [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+{
+  formatting = {
+    command = helpers.defaultNullOpts.mkListOf types.str "null" ''
+      External formatter command (with arguments).
+      It should accepts file content in stdin and print the formatted code into stdout.
+    '';
+  };
+
+  diagnostics = {
+    ignored = helpers.defaultNullOpts.mkListOf types.str [ ] ''
```
Comment (GaetanLepage) [inline plugins/lsp/language-servers/nil-ls-settings.nix:14]:
```suggestion
    ignored = helpers.defaultNullOpts.mkListOf types.str "[]" ''
```
Otherwise, this would translate in an empty default in the documentation as `toString [ ]` is `""`.

Comment (khaneliman) [reply plugins/lsp/language-servers/nil-ls-settings.nix]: Ah, makes sense. Forgot that was for documentation.

**Review by GaetanLepage — APPROVED** [review]

**Review by MattSturgeon — APPROVED** [review]

**Comment by GaetanLepage** [comment]
Thanks @khaneliman !
